### DISCUSSION TITLE
Getting events from child components

### DISCUSSION THREAD

**Discussion by serhiynovos** [arkivanov/Decompose#204] [Q&A] [answered]
Is it good idea to use callbacks to get events from child component ?

For example we have next components structure
- Root
  - Authorization
  - Main

Then for Root component I define next child factory

```kotlin
private fun createChild(
        config: Config,
        componentContext: ComponentContext
    ): Root.Child = when (config) {
        is Config.Authorization -> Root.Child.AuthorizationChild(
            component = AuthorizationComponent(
                componentContext = componentContext,
                storeFactory = storeFactory,
                mainContext = Dispatchers.Main,
                onAuthSuccess = {
                    // Replace authorization page by main page
                }
            )
        )
    
}
```

In my project I want to use coroutines, so maybe there are better ways how to achieve it ?

**Comment by arkivanov** [comment, marked as answer]
Yes it's a good idea. Should work just fine.

**Reply by serhiynovos** [reply]
But is we have multi nested structure and for example from 4th level i need to propagate event to root and to pass callback for every child

**Reply by arkivanov** [reply]
Yes, this makes callback dependencies explicit. This means that every subtree can handle the authentication.

**Reply by serhiynovos** [reply]
Ok. Thanks

**Reply by manosbatsis** [reply]
For nested structures it might be worth exploring the use of generic types and/or interface hierarchies for callback parameters? Perhaps enabling at least some callbacks in the component hierarchy to inspect/handle/propagate along with reasonable defaults  as appropriate.

**Reply by arkivanov** [reply]
@manosbatsis Do you have an example?

I used the following approach with delegation a couple of times, but then found that I like callbacks more.

```kotlin
class Level1(
    private val callbacks: Callbacks
) {
    fun level2(): Level2 =
        Level2(
            object : Level2.Callbacks, Callbacks by callbacks {
                override fun onFinished() {
                }

                override fun onItemSelected(id: Long) {
                }

                // onAuthRequired is delegated to callbacks
            }
        )

    interface Callbacks {
        fun onAuthRequired()
    }
}

class Level2(
    private val callbacks: Callbacks,
) {
    fun level3(): Level3 =
        Level3(
            object : Level3.Callbacks, Callbacks by callbacks {
                override fun onFinished() {
                    TODO("Not yet implemented")
                }
                
                // onAuthRequired is delegated to callbacks
            }
        )

    interface Callbacks {
        fun onFinished()
        fun onItemSelected(id: Long)
        fun onAuthRequired()
    }
}

class Level3(
    private val callbacks: Callbacks,
) {
    interface Callbacks {
        fun onFinished()
        fun onAuthRequired()
    }
}
```

**Reply by manosbatsis** [reply]
Ah, your pattern seems a good fit for my generated root-list-view-edit components. My suggestion was more like having a common interface between parent>child components to make a basic handler chain available. Only a low hanging fruit in my case really, mainly a lowest common denominator between cases like CRUD and wanting to reach some "root" component. For example:

```
interface GenericHandlerChainComponent{
   fun handle(input: Any) // or generic etc.
}
```
